src/Estimators/src/InvariantEKFBaseEstimator.cpp

```diff
@@ -0,0 +1,918 @@
+    stateStdDev.lContactFramePosition =  P.block<3, 3>(m_vecOffsets.lContactFramePosition, m_vecOffsets.lContactFramePosition).diagonal().array().sqrt();
+    if (estimateBias)
+    {
+        stateStdDev.gyroscopeBias =  P.block<3, 3>(m_vecOffsets.gyroBias, m_vecOffsets.gyroBias).diagonal().array().sqrt();
+        stateStdDev.accelerometerBias =  P.block<3, 3>(m_vecOffsets.accBias, m_vecOffsets.accBias).diagonal().array().sqrt();
+    }
+}
+
+void InvariantEKFBaseEstimator::Impl::constuctStateVar(const FloatingBaseEstimators::StateStdDev& stateStdDev,
+                                                       const bool& estimateBias,
+                                                       Eigen::MatrixXd& P)
+{
+    if (estimateBias)
+    {
+        P.resize(m_vecSizeWBias, m_vecSizeWBias);
+    }
+    else
+    {
+        P.resize(m_vecSizeWOBias, m_vecSizeWOBias);
+    }
+
+    P.setZero();
+    Eigen::Vector3d temp;
+
+    temp = stateStdDev.imuOrientation.array().square();
+    P.block<3, 3>(m_vecOffsets.imuOrientation, m_vecOffsets.imuOrientation) = temp.asDiagonal();
+    temp = stateStdDev.imuLinearVelocity.array().square();
+    P.block<3, 3>(m_vecOffsets.imuLinearVel, m_vecOffsets.imuLinearVel) = temp.asDiagonal();
+    temp = stateStdDev.imuPosition.array().square();
+    P.block<3, 3>(m_vecOffsets.imuPosition, m_vecOffsets.imuPosition) = temp.asDiagonal();
+    temp = stateStdDev.rContactFramePosition.array().square();
+    P.block<3, 3>(m_vecOffsets.rContactFramePosition, m_vecOffsets.rContactFramePosition) = temp.asDiagonal();
+    temp = stateStdDev.lContactFramePosition.array().square();
+    P.block<3, 3>(m_vecOffsets.lContactFramePosition, m_vecOffsets.lContactFramePosition) = temp.asDiagonal();
+
+    if (estimateBias)
+    {
+        temp = stateStdDev.gyroscopeBias.array().square();
+        P.block<3, 3>(m_vecOffsets.gyroBias, m_vecOffsets.gyroBias) = temp.asDiagonal();
+        temp = stateStdDev.accelerometerBias.array().square();
+        P.block<3, 3>(m_vecOffsets.accBias, m_vecOffsets.accBias) = temp.asDiagonal();
+    }
+}
+
+void InvariantEKFBaseEstimator::Impl::extractState(const FloatingBaseEstimators::InternalState& state,
+                                                   Eigen::MatrixXd& X,
+                                                   Eigen::Matrix<double, 6, 1>& theta)
+{
+    // X =    |R  v  p pr pl|
+    //        |   1         |
+    //        |      1      |
+    //        |         1   |
+    //        |            1|
+    X.resize(7, 7);
+    X.block<3, 3>(0, 0) = state.imuOrientation.toRotationMatrix();
+    X.block<3, 1>(0, 3) = state.imuLinearVelocity;
+    X.block<3, 1>(0, 4) = state.imuPosition;
+    X.block<3, 1>(0, 5) = state.rContactFramePosition;
+    X.block<3, 1>(0, 6) = state.lContactFramePosition;
+    X.bottomLeftCorner<4, 3>().setZero();
+    X.bottomRightCorner<4, 4>() = Eigen::Matrix4d::Identity();
```

Comment:
Probably it makes sense to store X as sparse matrix

Reply:
Given it's only a 7x7 matrix, do you think we can achieve higher gain margins using a sparse matrix?

Reply:
In our use cases, sparse matrices are useful only if need in some calculations involving inversions (e.g solving linear systems or when interfacing to solvers). Another possible application is to reduce the occupation in memory, but this is not an issue for basically all our applications. Hence, if we are not in the first case, I would avoid using sparse matrices.

Reply:
The most we do with this matrix is matrix multiplication. For inversions, we exploit with Lie group properties of direct product/semi-direct-product to only involve submatrix operations.

Reply:
Ok, I would vote for keeping it dense then.